builtin/logical/pki/acme_challenge_engine.go

```diff
@@ -58,33 +58,71 @@ func NewACMEChallengeEngine() *ACMEChallengeEngine {
 	return ace
 }
 
-func (ace *ACMEChallengeEngine) Initialize(b *backend, sc *storageContext) error {
-	if err := ace.LoadFromStorage(b, sc); err != nil {
-		return fmt.Errorf("failed loading initial in-progress validations: %w", err)
-	}
-
-	return nil
-}
-
-func (ace *ACMEChallengeEngine) LoadFromStorage(b *backend, sc *storageContext) error {
+func (ace *ACMEChallengeEngine) LoadFromStorage(sc *storageContext) error {
 	items, err := sc.Storage.List(sc.Context, acmeValidationPrefix)
 	if err != nil {
 		return fmt.Errorf("failed loading list of validations from disk: %w", err)
 	}
 
+	if len(items) == 0 {
+		return nil
+	}
+
 	ace.ValidationLock.Lock()
 	defer ace.ValidationLock.Unlock()
 
+	// Ensure we don't add any duplicate items to the queue.
+	existingItems := map[string]struct{}{}
+	element := ace.Validations.Front()
+	for element != nil {
+		value := element.Value.(*ChallengeQueueEntry)
+		existingItems[value.Identifier] = struct{}{}
+		element = element.Next()
+	}
+
 	// Add them to our queue of validations to work through later.
 	for _, item := range items {
+		if _, present := existingItems[item]; present {
+			continue
+		}
+
 		ace.Validations.PushBack(&ChallengeQueueEntry{
 			Identifier: item,
 		})
 	}
 
+	// Notify the running queue that there's new work to do.
+	select {
+	case ace.NewValidation <- items[0]:
+	default:
+	}
```

Comment:
```suggestion
	// Notify the running queue that there's new work to do.
	ace.NewValidation <- items[0]
```

Is there an advantage of using select here?

Reply:
@averche Is there an alternative to select that I'm missing to avoiding blocking?

<details>

<summary> <code>NewValidation</code> is a channel with a size of one: </summary>

https://github.com/hashicorp/vault/blob/d1c9318de52e15c5e5ff7f675045d6e46f533fcd/builtin/logical/pki/acme_challenge_engine.go#L53

This means that successive writes on this channel will block:

https://go.dev/play/p/JaUk2krI_iw

```
func main() {
	c := make(chan string, 1)
	c <- "one"
	c <- "two"   /// <<--- fatal error: all goroutines are asleep - deadlock!
	c <- "three"
	fmt.Println(<-c)
	fmt.Println(<-c)
	fmt.Println(<-c)
}
```

By using `select`, we can cancel the send without blocking, because there's already something in the queue that'll wake up the other goroutine. Note that this is just a notification that there's work available, the channel doesn't actually do anything with that information its given for now -- that's what the List is for.

In this example using `select`, we don't block on either reads or writes:

https://go.dev/play/p/sp_pQ5t2UcX

(same as above, just with selects with defaults)

outputs:

```
one
```

because only one write succeeded, only one read will succeed without blocking.

In reality, we probably want to add a timer case as well here:

https://github.com/hashicorp/vault/blob/d1c9318de52e15c5e5ff7f675045d6e46f533fcd/builtin/logical/pki/acme_challenge_engine.go#L112-L118

Just to ensure we do periodically trigger and release the per-goroutine channel resources, but it could be somewhat long (30s-5 minutes).

</details>

Reply:
Thank you very much for the detailed explanation 👍 That's a very interesting pattern that I haven't seen before!  Perhaps the comment(s) could be expanded with an abridged version of what you described here.